content/en/blog/_posts/2023-07-24-mixed-version-proxy-alpha/2023-07-24-mixed-version-proxy-alpha.md

```diff
@@ -0,0 +1,51 @@
+---
+layout: blog
+title: "Kubernetes 1.28: A New (alpha) Mechanism For Safer Cluster Upgrades"
+date: 2023-08-28
+slug: kubernetes-1-28-feature-mixed-version-proxy-alpha
+---
+
+**Author:** Richa Banker (Google)
+
+This blog describes the _mixed version proxy_, a new alpha feature in Kubernetes 1.28. The
+mixed version proxy enables an HTTP request for a resource to be served by the correct API server
+in cases where there are multiple API servers at varied versions in a cluster. For example,
+this is useful during a cluster upgrade, or when you're rolling out the runtime configuration of
+the cluster's control plane.
+
+## What problem does this solve?
+When a cluster undergoes an upgrade, the kube-apiservers existing at different versions in that scenario can serve different sets (groups, versions, resources) of built-in resources. A resource request made in this scenario may be served by any of the available apiservers, potentially resulting in the request ending up at an apiserver that may not be aware of the requested resource; consequently it being served a 404 not found error which is incorrect. Furthermore, incorrect serving of the 404 errors can lead to serious consequences such as namespace deletion being blocked incorrectly or objects being garbage collected mistakenly.
+
+## How do we solve the problem?
+
+{{< figure src="mvp-flow-diagram.png" caption="Mixed Version Proxy Flow Diagram" >}}
+
+The new feature “Mixed Version Proxy” provides the kube-apiserver with the capability to proxy a request to a peer kube-apiserver which is aware of the requested resource and hence can serve the request. To do this, a new filter has been added to the handler chain in the API server's aggregation layer.
+
+1. The new filter in the handler chain checks if the request is for a group/version/resource that the apiserver doesn't know about (using the existing [StorageVersion API](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/apis/apiserverinternal/types.go#L25-L37)). If so, it proxies the request to one of the apiservers that is listed in the ServerStorageVersion object. If the identified peer apiserver fails to respond (due to reasons like network connectivity, race between the request being received and the controller registering the apiserver-resource info in ServerStorageVersion object), then error 503("Service Unavailable") is served.
+2. To prevent indefinite rerouting of the request, a new header `X-Kubernetes-APIServer-Rerouted:true` is added to the original request once it is determined that the request cannot be served by the original apiserver that received the request and should therefore be proxied. The destination peer apiserver ensures to serve the request locally if it sees this header.
```

Comment:
```suggestion
2. To prevent indefinite proxying of the request, a (new for v1.28) HTTP header `X-Kubernetes-APIServer-Rerouted: true` is added to the original request once it is determined that the request cannot be served by the original API server. Setting that to true marks that the original API server couldn't handle the request and it should therefore be proxied. If a destination peer API server sees this header, it never proxies the request further.
```
(aside: does Kubernetes also set `Via:` headers?)

Reply:
Done. Sorry didn't quite understand the question. Only the new filter in the handler can/should set the new header.

Reply:
I think Kubernetes doesn't set `Via:` (lots of software doesn't, even though HTTP requires it)